src/pip/_internal/build_env.py

```diff
@@ -101,8 +106,49 @@ class SubprocessBuildEnvironmentInstaller:
     Install build dependencies by calling pip in a subprocess.
     """
 
-    def __init__(self, finder: PackageFinder) -> None:
+    def __init__(
+        self,
+        finder: PackageFinder,
+        build_constraints: list[str] | None = None,
+        build_constraint_feature_enabled: bool = False,
+        constraints: list[str] | None = None,
+    ) -> None:
         self.finder = finder
+        self._build_constraints = build_constraints or []
+        self._build_constraint_feature_enabled = build_constraint_feature_enabled
+        self._constraints = constraints or []
+
+    def _deprecation_constraint_check(self) -> None:
+        """
+        Check for deprecation warning: PIP_CONSTRAINT affecting build environments.
+
+        This warns when build-constraint feature is NOT enabled but regular constraints
+        match what PIP_CONSTRAINT environment variable points to.
+        """
+        if self._build_constraint_feature_enabled:
+            return
+
+        if not self._constraints:
+            return
+
+        if not os.environ.get("PIP_CONSTRAINT"):
+            return
+
+        pip_constraint_files = [
+            f.strip() for f in os.environ["PIP_CONSTRAINT"].split() if f.strip()
+        ]
+        if pip_constraint_files and set(pip_constraint_files) == set(self._constraints):
+            deprecated(
+                reason=(
+                    "Setting PIP_CONSTRAINT will not affect "
+                    "build constraints in the future,"
+                ),
+                replacement=(
+                    'PIP_BUILD_CONSTRAINT with PIP_USE_FEATURE="build-constraint"'
```

Comment:
Why do we have a two layer opt in?

Reply:
When you enable `PIP_USE_FEATURE="build-constraint"` it disables `PIP_CONSTRAINT` from affecting the build environment, regardless of whether you pass in build constraints or not.

To disable this warning you can enable the feature `build-constraint` but you don't have to actually pass in build constraints.

Open to simpler solutions that don't break existing behavior and warn users that might be affected.

Reply:
I suppose I could make passing in build constraints imply the feature `build-constraint` and then we can let the user know they can do either (pass in build constraints OR use the feature flag) to disable the warning?

Reply:
Yea, I think that'd be nicer than requiring two opt-ins.